File: blik/src/main/java/com/adyen/checkout/blik/BlikComponent.java
/*
 * Copyright (c) 2020 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by josephj on 4/12/2020.
 */

package com.adyen.checkout.blik;

import androidx.annotation.NonNull;
import androidx.lifecycle.SavedStateHandle;

import com.adyen.checkout.components.GenericComponentState;
import com.adyen.checkout.components.StoredPaymentComponentProvider;
import com.adyen.checkout.components.base.BasePaymentComponent;
import com.adyen.checkout.components.base.GenericPaymentMethodDelegate;
import com.adyen.checkout.components.base.GenericStoredPaymentComponentProvider;
import com.adyen.checkout.components.base.GenericStoredPaymentDelegate;
import com.adyen.checkout.components.model.payments.request.BlikPaymentMethod;
import com.adyen.checkout.components.model.payments.request.PaymentComponentData;
import com.adyen.checkout.components.util.PaymentMethodTypes;
import com.adyen.checkout.core.log.LogUtil;
import com.adyen.checkout.core.log.Logger;

public class BlikComponent extends BasePaymentComponent<BlikConfiguration, BlikInputData, BlikOutputData, GenericComponentState<BlikPaymentMethod>> {
    private static final String TAG = LogUtil.getTag();

    public static final StoredPaymentComponentProvider<BlikComponent, BlikConfiguration> PROVIDER =
            new GenericStoredPaymentComponentProvider<>(BlikComponent.class);

    private static final String[] PAYMENT_METHOD_TYPES = {PaymentMethodTypes.BLIK};

    public BlikComponent(
            @NonNull SavedStateHandle savedStateHandle,
            @NonNull GenericPaymentMethodDelegate paymentMethodDelegate,
            @NonNull BlikConfiguration configuration
    ) {
        super(savedStateHandle, paymentMethodDelegate, configuration);
    }

    public BlikComponent(
            @NonNull SavedStateHandle savedStateHandle,
            @NonNull GenericStoredPaymentDelegate paymentDelegate,
            @NonNull BlikConfiguration configuration
    ) {
        super(savedStateHandle, paymentDelegate, configuration);
        // TODO: 09/12/2020 move this logic to base component, maybe create the inputdata from the delegate?
        inputDataChanged(new BlikInputData());
    }

    @Override
    public boolean requiresInput() {
        return mPaymentMethodDelegate instanceof GenericPaymentMethodDelegate;
    }

    @NonNull
    @Override
    protected BlikOutputData onInputDataChanged(@NonNull BlikInputData inputData) {
        Logger.v(TAG, "onInputDataChanged");
        return new BlikOutputData(inputData.getBlikCode());
    }

    @NonNull
    @Override
    protected GenericComponentState<BlikPaymentMethod> createComponentState() {

        final BlikOutputData blikOutputData = getOutputData();

        final PaymentComponentData<BlikPaymentMethod> paymentComponentData = new PaymentComponentData<>();
        final BlikPaymentMethod paymentMethod = new BlikPaymentMethod();
        paymentMethod.setType(BlikPaymentMethod.PAYMENT_METHOD_TYPE);

        if (blikOutputData != null) {
            paymentMethod.setBlikCode(blikOutputData.getBlikCodeField().getValue());
        }

        if (mPaymentMethodDelegate instanceof GenericStoredPaymentDelegate) {
            paymentMethod.setStoredPaymentMethodId(
                    ((GenericStoredPaymentDelegate) mPaymentMethodDelegate).getStoredPaymentMethod().getId());
        }

        paymentComponentData.setPaymentMethod(paymentMethod);

        final boolean isInputValid = mPaymentMethodDelegate instanceof GenericStoredPaymentDelegate
                || blikOutputData != null
                && blikOutputData.isValid();

        return new GenericComponentState<>(paymentComponentData, isInputValid, true);
    }

    @NonNull
    @Override
    public String[] getSupportedPaymentMethodTypes() {
        return PAYMENT_METHOD_TYPES;
    }
}

File: blik/src/main/java/com/adyen/checkout/blik/BlikComponent.kt
/*
 * Copyright (c) 2020 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by josephj on 4/12/2020.
 */
package com.adyen.checkout.blik

import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.adyen.checkout.action.core.internal.ActionHandlingComponent
import com.adyen.checkout.action.core.internal.DefaultActionHandlingComponent
import com.adyen.checkout.action.core.internal.ui.GenericActionDelegate
import com.adyen.checkout.blik.internal.provider.BlikComponentProvider
import com.adyen.checkout.blik.internal.ui.BlikDelegate
import com.adyen.checkout.components.core.PaymentMethodTypes
import com.adyen.checkout.components.core.internal.ButtonComponent
import com.adyen.checkout.components.core.internal.ComponentEventHandler
import com.adyen.checkout.components.core.internal.PaymentComponent
import com.adyen.checkout.components.core.internal.PaymentComponentEvent
import com.adyen.checkout.components.core.internal.toActionCallback
import com.adyen.checkout.components.core.internal.ui.ComponentDelegate
import com.adyen.checkout.core.old.AdyenLogLevel
import com.adyen.checkout.core.old.internal.util.adyenLog
import com.adyen.checkout.ui.core.internal.ui.ButtonDelegate
import com.adyen.checkout.ui.core.internal.ui.ComponentViewType
import com.adyen.checkout.ui.core.internal.ui.ViewableComponent
import com.adyen.checkout.ui.core.old.internal.util.mergeViewFlows
import kotlinx.coroutines.flow.Flow

/**
 * A [PaymentComponent] that supports the [PaymentMethodTypes.BLIK] payment method.
 */
class BlikComponent internal constructor(
    private val blikDelegate: BlikDelegate,
    private val genericActionDelegate: GenericActionDelegate,
    private val actionHandlingComponent: DefaultActionHandlingComponent,
    internal val componentEventHandler: ComponentEventHandler<BlikComponentState>,
) : ViewModel(),
    PaymentComponent,
    ViewableComponent,
    ButtonComponent,
    ActionHandlingComponent by actionHandlingComponent {

    override val delegate: ComponentDelegate get() = actionHandlingComponent.activeDelegate

    override val viewFlow: Flow<ComponentViewType?> = mergeViewFlows(
        viewModelScope,
        blikDelegate.viewFlow,
        genericActionDelegate.viewFlow,
    )

    init {
        blikDelegate.initialize(viewModelScope)
        genericActionDelegate.initialize(viewModelScope)
        componentEventHandler.initialize(viewModelScope)
    }

    internal fun observe(
        lifecycleOwner: LifecycleOwner,
        callback: (PaymentComponentEvent<BlikComponentState>) -> Unit
    ) {
        blikDelegate.observe(lifecycleOwner, viewModelScope, callback)
        genericActionDelegate.observe(lifecycleOwner, viewModelScope, callback.toActionCallback())
    }

    internal fun removeObserver() {
        blikDelegate.removeObserver()
        genericActionDelegate.removeObserver()
    }

    override fun isConfirmationRequired(): Boolean = blikDelegate.isConfirmationRequired()

    override fun submit() {
        (delegate as? ButtonDelegate)?.onSubmit()
            ?: adyenLog(AdyenLogLevel.ERROR) { "Component is currently not submittable, ignoring." }
    }

    override fun setInteractionBlocked(isInteractionBlocked: Boolean) {
        (delegate as? BlikDelegate)?.setInteractionBlocked(isInteractionBlocked)
            ?: adyenLog(AdyenLogLevel.ERROR) { "Payment component is not interactable, ignoring." }
    }

    override fun onCleared() {
        super.onCleared()
        adyenLog(AdyenLogLevel.DEBUG) { "onCleared" }
        blikDelegate.onCleared()
        genericActionDelegate.onCleared()
        componentEventHandler.onCleared()
    }

    companion object {
        @JvmField
        val PROVIDER = BlikComponentProvider()

        @JvmField
        val PAYMENT_METHOD_TYPES = listOf(PaymentMethodTypes.BLIK)
    }
}
